### PR TITLE
Fix #79047 - Wasted stat call on workspace root 

### LINKED ISSUE #79047: explorer.autoReveal can produce a pointless stat call
Issue Type: <b>Bug</b>

Assume `explorer.autoReveal: true`

If a FileSystemProvider extension implements more than one scheme, and if a file is opened using a scheme that is not used by the root folder(s) of the workspace, then when focus is switched to that file a `stat()` call is issued on a folder resource that cannot possibly contain the file resource (because the scheme is different).

In some scenarios a stat() call on a root folder can be costly, meaning it is important not to do it unnecessarily.

This is not a new 1.37 issue.

I will add information about how to demonstrate the issue using a variant of fsprovider-sample, and I am also willing to supply a PR to fix it.

VS Code version: Code 1.37.0 (036a6b1d3ac84e5ca96a17a44e63a87971f8fcc8, 2019-08-08T02:33:50.993Z)
OS version: Windows_NT x64 10.0.17134

### PR DESCRIPTION
**Pull request by gjsjohnmurray**
This fix prevents a wasted stat call being made to a FileSystemProvider in specific circumstances. See #79047

### CODE REVIEW

File: src/vs/workbench/contrib/files/common/explorerService.ts [hunk 1]
```diff
@@ -159,6 +159,12 @@ export class ExplorerService implements IExplorerService {
 		const options: IResolveFileOptions = { resolveTo: [resource], resolveMetadata: this.sortOrder === 'modified' };
 		const workspaceFolder = this.contextService.getWorkspaceFolder(resource);
 		const rootUri = workspaceFolder ? workspaceFolder.uri : this.roots[0].resource;
+
```
Comment (isidorn) [inline src/vs/workbench/contrib/files/common/explorerService.ts]: Your check makes sense but I think we can make this more general.
I do not like the statment on line 161 that if the workspaceFolder does not exist we take the first root.

Let's change that, suche that if the workspaceFolder is not found the rootUri is undefined.
And if it is undefined we should just return.

I think this should also handle your use case. Let me know what you think about this and how ti goes.

**Review by isidorn — CHANGES_REQUESTED** [review]

**Comment by isidorn** [comment]
I have commented on your change directly. Once you address that we can merge this in.
The reason why I wouldp refer that approach is that if the resource does not belong to any root than a `select` is simply a no-op.

File: src/vs/workbench/contrib/files/common/explorerService.ts (hunk 1 above)
Comment (gjsjohnmurray) [reply src/vs/workbench/contrib/files/common/explorerService.ts]: @isidorn I agree. Have just updated my branch.

**Review by isidorn — APPROVED** [review]

**Comment by isidorn** [comment]
Looks great, merging in.
Thanks a lot!

**Pull request merged**